precompiles/authorization/ics20_approve.go

```diff
@@ -0,0 +1,264 @@
+	"github.com/ethereum/go-ethereum/accounts/abi"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/vm"
+	cmn "github.com/evmos/evmos/v14/precompiles/common"
+)
+
+// TransferMsg is the ICS20 transfer message type.
+var TransferMsg = sdk.MsgTypeURL(&transfertypes.MsgTransfer{})
+
+// Approve implements the ICS20 approve transactions.
+func Approve(
+	ctx sdk.Context,
+	authzKeeper authzkeeper.Keeper,
+	channelKeeper channelkeeper.Keeper,
+	precompileAddr, grantee, origin common.Address,
+	approvalExpiration time.Duration,
+	transferAuthz *transfertypes.TransferAuthorization,
+	event abi.Event,
+	stateDB vm.StateDB,
+) error {
+	// If one of the allocations contains a non-existing channel, throw and error
+	for _, allocation := range transferAuthz.Allocations {
+		found := channelKeeper.HasChannel(ctx, allocation.SourcePort, allocation.SourceChannel)
+		if !found {
+			return errorsmod.Wrapf(channeltypes.ErrChannelNotFound, "port ID (%s) channel ID (%s)", allocation.SourcePort, allocation.SourceChannel)
+		}
+	}
+
+	// Only the origin can approve a transfer to the grantee address
+	expiration := ctx.BlockTime().Add(approvalExpiration).UTC()
+	if err := authzKeeper.SaveGrant(ctx, grantee.Bytes(), origin.Bytes(), transferAuthz, &expiration); err != nil {
+		return err
+	}
+
+	allocations := convertToAllocation(transferAuthz.Allocations)
+	// Emit the IBC transfer authorization event
+	return EmitIBCTransferAuthorizationEvent(
+		event,
+		ctx,
+		stateDB,
+		precompileAddr,
+		grantee,
+		origin,
+		allocations,
+	)
+}
+
+// Revoke implements the ICS20 revoke transactions.
+func Revoke(
+	ctx sdk.Context,
+	authzKeeper authzkeeper.Keeper,
+	precompileAddr, grantee, origin common.Address,
+	event abi.Event,
+	stateDB vm.StateDB,
+) error {
+	// NOTE: we do not need to check the expiration as it will return nil if both found or expired
+	msgAuthz, _, err := CheckAuthzExists(ctx, authzKeeper, grantee, origin, TransferMsg)
+	if err != nil {
+		return fmt.Errorf(ErrAuthzDoesNotExistOrExpired, grantee, origin)
+	}
+
+	// check that the stored authorization matches the transfer authorization
+	if _, ok := msgAuthz.(*transfertypes.TransferAuthorization); !ok {
+		return authz.ErrUnknownAuthorizationType
+	}
+
+	if err = authzKeeper.DeleteGrant(ctx, grantee.Bytes(), origin.Bytes(), TransferMsg); err != nil {
+		return err
+	}
+
+	return EmitIBCRevokeAuthorizationEvent(
+		event,
+		ctx,
+		stateDB,
+		precompileAddr,
+		grantee,
+		origin,
+	)
+}
+
+// IncreaseAllowance implements the ICS20 increase allowance transactions.
+func IncreaseAllowance(
+	ctx sdk.Context,
+	authzKeeper authzkeeper.Keeper,
+	precompileAddr, grantee, origin common.Address,
+	sourcePort, sourceChannel, denom string,
+	amount *big.Int,
+	event abi.Event,
+	stateDB vm.StateDB,
+) error {
+	// NOTE: we do not need to check the expiration as it will return nil if both found or expired
```

Comment:
if found or expired or NOT found or expired?

Reply:
not found yes